contracts/src/v0.8/upkeeps/GovernanceUpkeep.sol

```diff
@@ -0,0 +1,137 @@
+  function checkUpkeep(
+    bytes calldata /* checkData */
+  ) external view override returns (bool upkeepNeeded, bytes memory performData) {
+    //Get number of proposal
+    uint256 proposalCount = s_governanceContract.proposalCount();
+
+    //Find starting index
+    uint256 newStartingIndex = findStartingIndex();
+
+    //If new starting index found, update in performUpkeep
+    if (newStartingIndex > s_proposalStartingIndex) {
+      performData = abi.encode(Action.UPDATE_INDEX, newStartingIndex);
+      return (true, performData);
+    }
+
+    //Go through each proposal and check the current state
+    for (uint256 i = newStartingIndex; i <= proposalCount; i++) {
+      GovernorAlpha.ProposalState state = s_governanceContract.state(i);
+      (, address proposer, , , , , , , ) = s_governanceContract.proposals(i);
+
+      if (state == GovernorAlpha.ProposalState.Succeeded) {
+        //If the state is 'Succeeded' then call 'queue' with the Proposal ID
+        performData = abi.encode(Action.QUEUE, i);
+        return (true, performData);
+      } else if (state == GovernorAlpha.ProposalState.Queued) {
+        //If the state is 'Queued' then call 'execute' with the Proposal ID
+        performData = abi.encode(Action.EXECUTE, i);
+        return (true, performData);
+      } else if (
+        s_governanceTokenContract.getPriorVotes(proposer, sub256(block.number, 1)) <
+        s_governanceContract.proposalThreshold()
+      ) {
+        performData = abi.encode(Action.CANCEL, i);
+        return (true, performData);
+      }
+    }
+
+    return (false, "");
+  }
+
+  ///@notice Chainlink Keepers will execute when checkUpkeep returns 'true'. Decodes the 'performData' passed in from checkUpkeep and performs an action as needed.
+  ///@param performData bytes encoded: (governance action required, index of proposal)
+  ///@dev The governance contract has action validation built-in
+  function performUpkeep(bytes calldata performData) external override {
+    //Decode performData
+    (Action performAction, uint256 proposalIndex) = abi.decode(performData, (Action, uint256));
+
+    //Check state of proposal at index
+    GovernorAlpha.ProposalState state = s_governanceContract.state(proposalIndex);
+
+    //Revalidate state and action of provided index
+    if (performAction == Action.QUEUE && state == GovernorAlpha.ProposalState.Succeeded) {
+      s_governanceContract.queue(proposalIndex);
+    } else if (performAction == Action.EXECUTE && state == GovernorAlpha.ProposalState.Queued) {
+      s_governanceContract.execute(proposalIndex);
+    } else if (performAction == Action.CANCEL && state != GovernorAlpha.ProposalState.Executed) {
+      s_governanceContract.cancel(proposalIndex);
+    } else if (performAction == Action.UPDATE_INDEX) {
+      uint256 newStartingIndex = findStartingIndex();
+      require(newStartingIndex > s_proposalStartingIndex, "No update required");
+      s_proposalStartingIndex = newStartingIndex;
+    }
```

Comment:
we should `revert()` if no matching action is found, otherwise nodes could send "useless" txs and user will get billed